velox/functions/sparksql/Comparisons.cpp

```diff
@@ -74,6 +92,77 @@ class ComparisonFunction final : public exec::VectorFunction {
       });
     }
   }
+
+ private:
+  template <typename T, bool constA, bool constB>
+  bool applySimdComparison(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      VectorPtr& result) const {
+    vector_size_t begin = rows.begin();
+    vector_size_t end = rows.end();
+    vector_size_t size = end - begin;
+    if (rows.countSelected() > 0.75 * size) {
+      const Cmp cmp;
+      tempRes_->reserve(rows.size());
+      auto __restrict tempRes = tempRes_->data();
+      auto __restrict rowsData = rows.allBits();
+      auto resultVector = result->asUnchecked<FlatVector<bool>>();
+      auto rawResult = resultVector->mutableRawValues<uint8_t>();
+      if constexpr (!constA && !constB) {
+        T* __restrict rawA =
+            args[0]->asUnchecked<FlatVector<T>>()->mutableRawValues();
+        T* __restrict rawB =
+            args[1]->asUnchecked<FlatVector<T>>()->mutableRawValues();
+        for (auto i = begin; i < end; i++) {
+          tempRes[i] =
```

Comment:
The Presto version of comparisons are more efficient by process one SIMD register at a time and pack the result, instead of maintain an extra buffer.  Do you consider apply the same approach?  Maybe some code can even be reused.

Reply:
Thank you for pointing this out. Yes, I referenced the implementation of Preso version comparisons and used some logic here to convert the result buffer into a result vector.
The problem I attempted to solve by introducing the extra buffer is that the result is set per bit, but we need to load a uint8_t value to update the result. Sometimes, the same uint8_t value is loaded and updated sametime by vectorization, leading to an incorrect final result. Here, as you see, I am just using a for loop to wrap the original comparison logic, as I couldn’t find a better way to ensure the result is correctly updated by auto-vectorization generated SIMD. If you have any good suggestions for this, please let me know.

Reply:
Yeah I mean you can use xsimd like we did in https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/Comparisons.cpp

Reply:
I agree why not use the Presto version of comparisons directly @zhli1142015 ,gluten only need to do is to change the function map .https://github.com/apache/incubator-gluten/discussions/6247

> Yeah I mean you can use xsimd like we did in https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/Comparisons.cpp

Reply:
I think we can use Presto version for tinyint, smallint, int and bigint. For float and double, Presto and Spark have different semantics, related to Nan handling. Before we tried to write SIMD for them but it looks very complex.
And as I explained in https://github.com/facebookincubator/velox/issues/10277#issuecomment-2184986474, we are looking to cover more types, like timstamp and decimal. And they are not covered by Presto version functions.

Reply:
> I think we can use Presto version only for tinyint, smallint, int and bigint. For float and double, Presto and Spark have different semantics, related to Nan handling. Before we tried to implement them with SMD but it looks very complex. And as I explained in [#10277 (comment)](https://github.com/facebookincubator/velox/issues/10277#issuecomment-2184986474), we are looking to cover more types, like timstamp and decimal. And they are not covered by Presto version functions.

thanks a lot.I got it. if possible，I can do some work on the imporvement.

Reply:
> For float and double, Presto and Spark have different semantics, related to Nan handling.

I notice there is some on-going work to fix the NaN semantics of Presto. Would the difference be resolved in the future?

Reply:
Seems so.

Reply:
Auto-vectorization is only responsible for optimizing data comparisons performance. I actually referred to Presto's SIMD implementation for the logic of result updates and used a batch size of 32 for all data types. I split the logic into two parts: data comparison and result updating by introducing temporary vectors to store comparison results. I think there are two benefits: we can obtain performance gains from the result update part even if the data type is not supported by auto-vectorization. If it’s a data type and comparison logic that can be converted by auto-vectorization, we can achieve performance gains for data comparison through auto-vectorization without needing to maintain separate SIMD code. Please let me know if this makes sense to you.

Reply:
> > For float and double, Presto and Spark have different semantics, related to Nan handling.
> 
> I notice there is some on-going work to fix the NaN semantics of Presto. Would the difference be resolved in the future?

@rui-mo , should be yes. I saw this PR merged for Presto: https://github.com/facebookincubator/velox/pull/10165

Reply:
@zhli1142015 Maintaining temporary result bools makes some types slower.  Since we already implemented it on Presto, we can just reuse the code with minor refactor, there will be no extra code to maintain.  Auto vectorization can be added for types that does not support SIMD (e.g. long decimal).

Reply:
Sure, I extract `applySimdComparison` to SIMDComparisionUtil.h file also and use it for both Presto and Spark, please help check, thanks.